consensus/istanbul/uptime/uptime.go

```diff
@@ -0,0 +1,242 @@
+package uptime
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+// Store provides a persistent storage for uptime entries
+type Store interface {
+	ReadAccumulatedEpochUptime(epoch uint64) *Uptime
+	WriteAccumulatedEpochUptime(epoch uint64, uptime *Uptime)
+}
+
+// Uptime contains the latest block for which uptime metrics were accounted for. It also contains
+// an array of Entries where the `i`th entry represents the uptime statistics of the `i`th validator
+// in the validator set for that epoch
+type Uptime struct {
+	LatestBlock uint64
+	Entries     []UptimeEntry
+}
+
+// UptimeEntry contains the uptime score of a validator during an epoch as well as the
+// last block they signed on
+type UptimeEntry struct {
+	ScoreTally      uint64
+	LastSignedBlock uint64
+}
+
+func (u *UptimeEntry) String() string {
+	return fmt.Sprintf("UptimeEntry { scoreTally: %v, lastBlock: %v}", u.ScoreTally, u.LastSignedBlock)
+}
+
+// GetUptimeMonitoringWindow retrieves the range [first block, last block] where uptime is to be monitored
+// for a give epoch. The range is inclusive.
+// First blocks of an epoch need to be skipped since we can't assess the last `lookbackWindow` block for validators
+// as those are froma different epoch.
```

Comment:
this might be clearer with something like: "we do not monitor while the lookback window crosses the epoch boundary."

Reply:
Changed to:
```
// We do not monitor while the lookback window crosses the epoch boundary; since we can only analyze window when the window
// (block range) belongs to a single block (thus initial epoch blocks are skipped)
```

Reply:
End up removing the comment, in favor of inline comments in the function. Check @oneeman comment